ibis/spark/compiler.py

```diff
@@ -26,13 +68,356 @@ class SparkContext(ImpalaContext):
+    op = expr.op()
+    arg, digits = op.args
+
+    arg_formatted = translator.translate(arg)
+
+    if digits is not None:
+        digits_formatted = translator.translate(digits)
+        return 'bround({}, {})'.format(arg_formatted, digits_formatted)
+    return 'bround({})'.format(arg_formatted)
+
+
+_operation_registry = impala_compiler._operation_registry.copy()
+_operation_registry.update(
+    {
+        ops.IsNan: unary('isnan'),
+        ops.IsInf: _is_inf,
+        ops.IfNull: fixed_arity('ifnull', 2),
+        ops.StructField: _struct_field,
+        ops.MapValueForKey: _map_value_for_key,
+        ops.ArrayLength: unary('size'),
+        ops.Round: _round,
+        ops.HLLCardinality: _reduction('approx_count_distinct'),
+        ops.StrRight: fixed_arity('right', 2),
+        ops.StringSplit: fixed_arity('SPLIT', 2),
+        ops.RegexSearch: fixed_arity('rlike', 2),
+        ops.StringConcat: _string_concat,
+        ops.ArrayConcat: fixed_arity('concat', 2),
+        ops.Cast: _cast,
+        ops.ExtractYear: unary('year'),
+        ops.ExtractMonth: unary('month'),
+        ops.ExtractDay: unary('day'),
+        ops.ExtractHour: unary('hour'),
+        ops.ExtractMinute: unary('minute'),
+        ops.ExtractSecond: unary('second'),
+        ops.DateAdd: _timestamp_op('date_add'),
+        ops.DateSub: _timestamp_op('date_sub'),
+        ops.TimestampAdd: _timestamp_op('date_add'),
+        ops.TimestampSub: _timestamp_op('date_sub'),
+        ops.TimestampDiff: _timestamp_diff,
+        ops.TimestampTruncate: _timestamp_truncate,
+        ops.TimestampFromUNIX: _timestamp_from_unix,
+        ops.DateTruncate: _date_truncate,
+        ops.Literal: _literal,
+    }
+)
+
+
 class SparkExprTranslator(ImpalaExprTranslator):
+    _registry = _operation_registry
+
     context_class = SparkContext
 
 
+compiles = SparkExprTranslator.compiles
+rewrites = SparkExprTranslator.rewrites
+
+
+@compiles(ops.Arbitrary)
+def spark_compiles_arbitrary(translator, expr):
+    arg, how, where = expr.op().args
+
+    if where is not None:
+        arg = where.ifelse(arg, ibis.NA)
+
+    if how in (None, 'first'):
+        return 'first({}, True)'.format(translator.translate(arg))
+    elif how == 'last':
+        return 'last({}, True)'.format(translator.translate(arg))
+    else:
+        raise com.UnsupportedOperationError(
+            '{!r} value not supported for arbitrary in Spark SQL'.format(how)
+        )
+
+
+@compiles(ops.Strftime)
+def spark_compiles_strftime(translator, expr):
+    arg, format_string = expr.op().args
+
+    return 'date_format({}, {})'.format(
+        translator.translate(arg),
+        translator.translate(format_string)
+    )
+
+
+@rewrites(ops.DayOfWeekName)
+def spark_rewrites_day_of_week_name(expr):
+    arg = expr.op().arg
+    # TODO: if strftime gets rewritten to be compatible with ibis standard
+    # strftime, then 'EEEEE' needs to be rewritten to '%A'
+    return arg.strftime('EEEEE')
```

Comment:
Is this really how to specify the name of the day of week??

Reply:
Turns out just four Es suffices, major improvement